File: python/7_kyu/cryptanalysis_word_patterns.py
# Cryptanalysis Word Patterns
# https://www.codewars.com/kata/5f3142b3a28d9b002ef58f5e/train/python

def word_pattern(word):
    # create a count variable to hold current indices
    indices = 0
    # create an empty dictionary to hold letter:indices values
    crypto = {}

    # iterate through the word and append unique letters and their corresponding indices
    for i in word:
        if i.lower() not in crypto:
            crypto[i.lower()] = str(indices)
            indices += 1

    # create an empty array to hold the indices before being joined by .
    ans = []

    # iterate through the word again and append the value corresponding to the letter from the dictionary
    for i in word:
        ans.append(crypto[i.lower()])
    
    # return the joined version of the answer array
    return '.'.join(ans)

    

print(word_pattern("Hippopotomonstrosesquippedaliophobia"))

File: javascript/7_kyu/cryptanalysis_word_patterns.js
// Cryptanalysis Word Patterns
// https://www.codewars.com/kata/5f3142b3a28d9b002ef58f5e/train/javascript

function wordPattern(word) {
    // Create a variable that will hold current indices
    var indices = 0;
    // create a dictionary that will hold letter:indices
    var letters = {};
    // iterate through the word and append the letter to number
    for (var i = 0; i < word.length; i++) {
        if ( !(word[i].toLowerCase() in letters) ) {
            letters[word[i].toLowerCase()] = indices.toString();
            indices++;
        }
    }
    // create an empty array to hold the indices comma separated
    var ans = [];

    // iterate through the word again, pushing letter's indices into the empyt array
    for (var i = 0; i < word.length; i++) {
        ans.push(letters[word[i].toLowerCase()]);
    }
    // return the joined version of the answer array
    return ans.join('.');
}

console.log(wordPattern("Hippopotomonstrosesquippedaliophobia"));